eng/common/scripts/job-matrix/Create-PrJobMatrix.ps1

```diff
@@ -0,0 +1,129 @@
+<#
+.SYNOPSIS
+Generates a combined PR job matrix from a package properties folder. It is effectively a combination of
+Create-JobMatrix and distribute-packages-to-matrix.
+
+.DESCRIPTION
+Create-JobMatrix has a limitation in that it accepts one or multiple matrix files, but it doesn't allow runtime
+selection of the matrix file based on what is being built. Due to this, this script exists to provide exactly
+that mapping.
+
+It should be called from a PR build only.
+
+It generates the matrix by the following algorithm:
+  - load all package properties files
+  - group the package properties by their targeted CI Matrix Configs
+    - for each package group, generate the matrix for each matrix config in the group (remember MatrixConfigs is a list not a single object)
+      - for each matrix config, generate the matrix
+        - calculate if batching is necessary for this matrix config
+        - for each batch
+          - create combined property name for the batch
+          - walk each matrix item
+            - add suffixes for batch and matrix config if nececessary to the job name
+            - add the combined property name to the parameters of the matrix item
+            - add the matrix item to the overall result
+
+.EXAMPLE
+./eng/common/scripts/job-matrix/Create-PrJobMatrix.ps1 `
+  -PackagePropertiesFolder "path/to/populated/PackageInfo" `
+  -PrMatrixSetting "<Name of variable to set in the matrix>"
+#>
+
+[CmdletBinding()]
+param (
+  [Parameter(Mandatory = $true)][string] $PackagePropertiesFolder,
+  [Parameter(Mandatory = $true)][string] $PRMatrixFile,
+  [Parameter(Mandatory = $true)][string] $PRMatrixSetting,
+  [Parameter(Mandatory = $False)][string] $DisplayNameFilter,
+  [Parameter(Mandatory = $False)][array] $Filters,
+  [Parameter(Mandatory = $False)][array] $Replace,
+  [Parameter()][switch] $CI = ($null -ne $env:SYSTEM_TEAMPROJECTID)
+)
+
+. $PSScriptRoot/job-matrix-functions.ps1
+. $PSScriptRoot/../Helpers/Package-Helpers.ps1
+$BATCHSIZE = 10
+
+if (!(Test-Path $PackagePropertiesFolder)) {
+  Write-Error "Package Properties folder doesn't exist"
+  exit 1
+}
+
+if (!(Test-Path $PRMatrixFile)) {
+  Write-Error "PR Matrix file doesn't exist"
+  exit 1
+}
+
+Write-Host "Generating PR job matrix for $PackagePropertiesFolder"
+
+$configs = Get-Content -Raw $PRMatrixFile | ConvertFrom-Json | ForEach-Object { [PSCustomObject]$_ }
+
+# calculate general targeting information and create our batches prior to generating any matrix
+$packageProperties = Get-ChildItem -Recurse "$PackagePropertiesFolder" *.json `
+| ForEach-Object { Get-Content -Path $_.FullName | ConvertFrom-Json } `
+| ForEach-Object { [PSCustomObject]$_ }
+
+# set default matrix config for each package if there isn't an override
+$packageProperties | ForEach-Object {
+  if (-not $_.CIMatrixConfigs) {
+    $_.CIMatrixConfigs = $configs
+  }
```

Comment:
If this is only null and not empty string you can simplify to:

```suggestion
  $_.CIMatrixConfigs = $_.CIMatrixConfigs ?? $configs
```

Reply:
Will wait for the rest of your comments and pull it in.